packages/token/src/tests/erc20/test_erc20_permit.cairo

```diff
@@ -0,0 +1,487 @@
+use core::hash::{HashStateTrait, HashStateExTrait};
+use core::poseidon::PoseidonTrait;
+use crate::erc20::ERC20Component::{ERC20MixinImpl, InternalImpl};
+use crate::erc20::extensions::erc20_permit::{Permit, PERMIT_TYPE_HASH};
+use crate::erc20::interface::{ERC20PermitABIDispatcher, ERC20PermitABIDispatcherTrait};
+use crate::tests::mocks::erc20_permit_mocks::DualCaseERC20PermitMock;
+use openzeppelin_testing as utils;
+use openzeppelin_testing::constants;
+use openzeppelin_testing::signing::{StarkKeyPair, StarkSerializedSigning};
+use openzeppelin_utils::cryptography::snip12::{StarknetDomain, StructHashStarknetDomainImpl};
+use openzeppelin_utils::serde::SerializedAppend;
+use snforge_std::signature::stark_curve::StarkCurveSignerImpl;
+use snforge_std::{
+    start_cheat_caller_address, start_cheat_block_timestamp, start_cheat_chain_id_global
+};
+use starknet::ContractAddress;
+
+//
+// Test Data
+//
+
+#[derive(Copy, Drop)]
+struct TestData {
+    contract_address: ContractAddress,
+    owner: ContractAddress,
+    key_pair: StarkKeyPair,
+    spender: ContractAddress,
+    amount: u256,
+    deadline: u64,
+    token_supply: u256,
+    name: @ByteArray,
+    symbol: @ByteArray,
+    metadata_name: felt252,
+    metadata_version: felt252,
+    chain_id: felt252,
+    revision: felt252
+}
+
+fn TEST_DATA() -> TestData {
+    TestData {
+        contract_address: constants::CONTRACT_ADDRESS(),
+        owner: constants::OWNER(),
+        key_pair: constants::stark::KEY_PAIR(),
+        spender: constants::SPENDER(),
+        amount: constants::TOKEN_VALUE,
+        deadline: constants::TIMESTAMP,
+        token_supply: constants::SUPPLY,
+        name: @constants::NAME(),
+        symbol: @constants::SYMBOL(),
+        metadata_name: DualCaseERC20PermitMock::SNIP12MetadataImpl::name(),
+        metadata_version: DualCaseERC20PermitMock::SNIP12MetadataImpl::version(),
+        chain_id: constants::CHAIN_ID,
+        revision: 1 // As in the current SNIP-12 implementation
+    }
+}
+
+//
+// Setup
+//
+
+fn setup(data: TestData) -> ERC20PermitABIDispatcher {
+    start_cheat_chain_id_global(data.chain_id);
+
+    utils::declare_and_deploy_at(
+        "DualCaseAccountMock", data.owner, array![data.key_pair.public_key]
+    );
```

Comment:
Is something preventing us from using `COMPONENT_STATE` instead of deploying and dispatchers? Note that using COMPONENT_STATE should speed up running the tests.

Reply:
There's nothing preventing us from following the approach with `CONTRACT_STATE`, but there's also no strong reason why this approach is better. I reckon that having a contract deployed and interacting with it through dispatcher feels closer to the real-world usage scenario and also prevents from using any internal functions that 
are not supposed and even can't be called externally

Either way tests run pretty fast. That's compilation that takes most time

Reply:
I had the same thoughts when we started following this logic, we already had a discussion about it and decided to test components as separate modules, since they are not deployed contracts, but modules to extend those. Declaring and deploying in tests take time and adds up while we continue adding more tests. I'm not against reopening the discussion about it, but for now, since we should always favor consistency, let's refactor into using COMPONENT_STATE as we've been doing when possible for component tests.

Reply:
Refactored to use component state instead of deploying a mock